File: Assets/Scripts/GameCommon/LuaTimer.cs
using System.Collections.Generic;
using UnityEngine;

public class LuaTimer : MonoBehaviour
{
    public string scriptName = "Common/LuaTimer";
    private ScriptInstaller lua = new ScriptInstaller();
    private string moduleName;

    public delegate void OnTimer();
    private Dictionary<long, Element> mElements = new Dictionary<long, Element>();
    private List<long> mTempDel = new List<long>(); 

    private long StartID = 0;

    private class Element
    {
        public long mId;
        public float mElasped = 0f;
        public bool mIgnoreTimeScale = true;
        public float mDuration = 0f;
        public OnTimer mOnTimer;

		public bool OnTimePassed(float deltaTime)
        {
            var dur = mDuration;
		    if (mDuration < 0)
		    {
		        dur = -mDuration;
		    }

		    mElasped = mElasped + deltaTime;

			if (mElasped >= dur)
            {
                if (mOnTimer != null)
                {
                    mOnTimer();
                }

                if (mDuration > 0)
                {
                    return true;
                }
                else
                {
                    mElasped = mElasped - dur;
                    return false;
                }
            }
			else
			{
                return false;
			}
		}
    }

    public static LuaTimer Instance;

    void Awake()
    {
        if (Instance == null)
        {
            Instance = this;
        }
        else
        {
            LogView.Error("LuaTimer is singleton!");
            Destroy(gameObject);
        }
    }

    void OnDestroy()
    {
        if (Instance == this)
        {
            Instance = null;
        }
    }

    void Start()
    {
        lua.Install(LuaScriptMgr.Instance, Util.GetLuaFilePath(scriptName));
        moduleName = Util.GetLuaScriptModuleName(scriptName);
    }

	// Update is called once per frame
	void Update ()
    {
        LuaScriptMgr.Instance.CallLuaFunction(moduleName + "Update", Time.time, Time.realtimeSinceStartup);

	    foreach (var kvp in mElements)
	    {
	        var delta = kvp.Value.mIgnoreTimeScale ? RealTime.deltaTime : RealTime.time;
	        if (kvp.Value.OnTimePassed(delta))
	        {
	            mTempDel.Add(kvp.Key);
	        }
	    }

	    foreach (var id in mTempDel)
	    {
	        mElements.Remove(id);
	    }
	    mTempDel.Clear();
    }

    public long AddTimer(float dur, bool ignoreTimeScale, OnTimer onTimer)
    {
        var e = new Element { mId = NextID(), mDuration = dur, mIgnoreTimeScale = ignoreTimeScale, mOnTimer = onTimer };
        mElements.Add(e.mId, e);
        return e.mId;
    }

    public void RmvTimer(long id)
    {
        mElements.Remove(id);
    }

    private long NextID()
    {
        return ++StartID;
    }
}

File: Assets/Lua/Common/LuaTimer.lua
module("LuaTimer", package.seeall)
require('Common/TimerObj')

local activeTimers = {};
setmetatable(activeTimers, {__mode = "k"});

local timerId = 0;
local mTime = 0;	--Time.time
local mRealTimeSinceStartup = 0; --Time.realTimeSinceStart


function Update(time, realTimeSinceStartup)
	
	local delta = time - mTime;
	local realDelta = realTimeSinceStartup - mRealTimeSinceStartup;

	for k, v in pairs(activeTimers) do
		local rmv = false;
		if (v.mIgnoreTimeScale) then
			rmv = v:OnTimePassed(realDelta);
		else
			rmv = v:OnTimePassed(delta);
		end

		if rmv then
			activeTimers[k] = nil;
		end
	end

	
	mTime = time;
	mRealTimeSinceStartup = realTimeSinceStartup;
end

function AddTimer(ignoreTimeScale, duration, OnTimer)
	timerId = tostring(timerId+1);
	newTimer = TimerObj.new(timerId, ignoreTimeScale, duration, OnTimer);
	activeTimers[timerId] = newTimer;
	return timerId;
end

function RmvTimer(id)
    activeTimers[id] = nil;
end